### DISCUSSION TITLE
app.acquire_token_for_client error - AssertionError: Invalid parameter type

### DISCUSSION THREAD

**Discussion by Git-BT-Hub** [AzureAD/microsoft-authentication-library-for-python#587] [Q&A] [answered]
Getting Error while trying to acquire Token using below 

from msal import ConfidentialClientApplication

#import secrets from keyvault
service_principal_id = dbutils.secrets.get(scope = secret_scope, key = "clientid")
service_principal_secret = dbutils.secrets.get(scope = secret_scope, key = "clientsecret")
tenant_id = dbutils.secrets.get(scope = secret_scope, key = "tenant-id")
authority = "https://login.windows.net/" + tenant_id

app = ConfidentialClientApplication(
    client_id=service_principal_id,
    client_credential=service_principal_secret,
    authority=authority
)
s = "https://graph.microsoft.com/.default"
**result = app.acquire_token_for_client(scopes=s)**
a = result.get("access_token")
print(a)

Getting Error on this line , this was working fine until 21-Jul-2023 
app.acquire_token_for_client 

Please let me know what has changed in MSAL(PYPI) or what is the alternative with latest version 

If this can work with Previous Version let me know, which version to use 

Thanks

**Comment by JayCork** [comment]
I think that the logic from [this commit ](https://github.com/AzureAD/microsoft-authentication-library-for-python/commit/1b316e31ecc8c10f5c95f8285ffd50d45ea1c1d2) changed scope typing.

**Comment by rayluo** [comment]
> Getting Error on this line, *this was working fine until 21-Jul-2023*
> app.acquire_token_for_client

Out of curiosity, does your environment somehow always pull in the latest MSAL version, *unknowingly*? Most bug reports would pinpoint the differences as either "this was working until I upgrade to your new version X.Y.Z" or "this was working until I made a change in the code snippet" etc..

That being said, lucky that in this case, yes we can correlate the timing with our recent release of MSAL 1.23. Indeed, the following line works with MSAL 1.22 but can repro your error in MSAL 1.23.

    python -c "from msal import ConfidentialClientApplication as CCA; print(CCA('id', client_credential='foo').acquire_token_for_client('https://graph.microsoft.com/.default'))" 

We shall probably improve that exception message in the future. Meanwhile, as hinted by the line number in that exception, the error was about the scope input was meant to be a list, not a string.

**If at all possible, please change your `"scope"` to `["scope"]`, and then stay with latest MSAL 1.23.**

By the way, is your app a long-lived daemon, or a one-time command-line tool? If it is the latter, you shall further consider to [enable a token cache](https://msal-python.readthedocs.io/en/latest/#msal.SerializableTokenCache), otherwise your app risks being throttled if/when it sends out too much token requests.

**Comment by p17b** [comment, marked as answer]
had the same issue with release 1.23.0, was ok in 1.22.0
I found that the 'scopes' parameter should be a list of strings.
as @rayluo said tried to wrap in [] 
eg.
` s = ["https://graph.microsoft.com/.default"]`

**Reply by Raghavendra-Dasara** [reply]
Thank you :)

**Reply by rayluo** [reply]
For those who are still running into this old topic, you probably encountered this exception which hinted that the input shall be a list (of strings).

```
    assert isinstance(scopes, list), "Invalid parameter type"
AssertionError: Invalid parameter type
```

**Comment by Git-BT-Hub** [comment]
> > Getting Error on this line, _this was working fine until 21-Jul-2023_
> > app.acquire_token_for_client
> 
> Out of curiosity, does your environment somehow always pull in the latest MSAL version, _unknowingly_? Most bug reports would pinpoint the differences as either "this was working until I upgrade to your new version X.Y.Z" or "this was working until I made a change in the code snippet" etc..
> 
> That being said, lucky that in this case, yes we can correlate the timing with our recent release of MSAL 1.23. Indeed, the following line works with MSAL 1.22 but can repro your error in MSAL 1.23.
> 
> ```
> python -c "from msal import ConfidentialClientApplication as CCA; print(CCA('id', client_credential='foo').acquire_token_for_client('https://graph.microsoft.com/.default'))" 
> ```
> 
> We shall probably improve that exception message in the future. Meanwhile, as hinted by the line number in that exception, the error was about the scope input was meant to be a list, not a string.
> 
> If at all possible, please change your `"scope"` to `["scope"]`, and then stay with latest MSAL 1.23.
> 
> By the way, is your app a long-lived daemon, or a one-time command-line tool? If it is the latter, you shall further consider to [enable a token cache](https://msal-python.readthedocs.io/en/latest/#msal.SerializableTokenCache), otherwise your app risks being throttled if/when it sends out too much token requests.

Hi @rayluo - Thats correct, I always get latest version on MSAL, but after this error when I switched to v1.22, it seems to be working fine . I will try converting "scope" to ["scope"], and see if that works for me. [I am using Azure Databricks]

Thanks

**Comment by syntrixsystems** [comment]
Myself and a colleague both have this issue at 2 different clients as of yesterday.

My colleague noted just now that if you enclose the scopes value in brackets, seems to have resolved the issue for us.  scopes must be expecting a list now?

token = app.acquire_token_for_client(scopes=**[**"https://database.windows.net/.default"**]**)["access_token"]

**Comment by rayluo** [comment]
> My colleague noted just now that **if you enclose the scopes value in brackets, seems to have resolved the issue for us. scopes must be expecting a list now?**
> 
> token = app.acquire_token_for_client(scopes=**[**"https://database.windows.net/.default"**]**)["access_token"]

The `scopes` parameter was meant to be a list of strings, as all other acquire token APIs do. [Our docs said so for the last 5 years](https://github.com/AzureAD/microsoft-authentication-library-for-python/blame/dev/msal/application.py#L2033-L2034). Looks like `acquire_token_for_client()` accidentally allowed string.

> Myself and a colleague both have this issue at 2 different clients as of yesterday.
> 
Would you be able to ship your `["scope"]` fix with/to your clients? @syntrixsystems

**Comment by mkemmerz** [comment]
Will the parameter stay a list now or will the change be reverted?

**Comment by syntrixsystems** [comment]
> > My colleague noted just now that **if you enclose the scopes value in brackets, seems to have resolved the issue for us. scopes must be expecting a list now?**
> > token = app.acquire_token_for_client(scopes=**[**"https://database.windows.net/.default"**]**)["access_token"]
> 
> The `scopes` parameter was meant to be a list of strings, as all other acquire token APIs do. [Our docs said so for the last 5 years](https://github.com/AzureAD/microsoft-authentication-library-for-python/blame/dev/msal/application.py#L2033-L2034). Looks like `acquire_token_for_client()` accidentally allowed string.
> 
> > Myself and a colleague both have this issue at 2 different clients as of yesterday.
> 
> Would you be able to ship your `["scope"]` fix with/to your clients? @syntrixsystems

We both implemented the change in our respective code bases and have had no issues since.

**Comment by rayluo** [comment]
> Will the parameter stay a list now or will the change be reverted?

The `scopes` parameter has always been a plural, so it is meant to accept a list of strings. This part will never change.

Even if we might bring back the "scopes as a string" usage, it will not be a revert. It will be with a [DeprecationWarning](https://docs.python.org/3/library/exceptions.html#DeprecationWarning). So, you are still recommended to switch to the list-of-string syntax, in order to be future proof.

**Comment by disciple-dev** [comment]
> Out of curiosity, does your environment somehow always pull in the latest MSAL version, _unknowingly_? Most bug reports would pinpoint the differences as either "this was working until I upgrade to your new version X.Y.Z" or "this was working until I made a change in the code snippet" etc..

Auto upgrades aside, it is evident that this change was not backwards-compatible, and introduced a breaking change.

**Comment by rayluo** [comment]
> it is evident that this change was not backwards-compatible, and introduced a breaking change.

To set the record straight, the concept of backwards-compatible or breaking-change does not apply to implementation details. In this case, [the docs](https://github.com/AzureAD/microsoft-authentication-library-for-python/blame/dev/msal/application.py#L2033-L2034) and [the sample](https://github.com/AzureAD/microsoft-authentication-library-for-python/blame/dev/sample/confidential_client_secret_sample.py#L7) have all been a list of strings for 4 or 5 years. There is also [the 2nd bullet point here](https://msal-python.readthedocs.io/en/latest/#api-reference) saying that the undocumented behaviors are subject to change without prior notice.

**Comment by disciple-dev** [comment]
> > it is evident that this change was not backwards-compatible, and introduced a breaking change.
> 
> To set the record straight, the concept of backwards-compatible or breaking-change does not apply to implementation details. In this case, [the docs](https://github.com/AzureAD/microsoft-authentication-library-for-python/blame/dev/msal/application.py#L2033-L2034) and [the sample](https://github.com/AzureAD/microsoft-authentication-library-for-python/blame/dev/sample/confidential_client_secret_sample.py#L7) have all been a list of strings for 4 or 5 years. There is also [the 2nd bullet point here](https://msal-python.readthedocs.io/en/latest/#api-reference) saying that the undocumented behaviors are subject to change without prior notice.

I stand corrected; my apologies.
